src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryCopy.java

```diff
@@ -0,0 +1,574 @@
+/*
+ *  Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ *  This code is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 only, as
+ *  published by the Free Software Foundation.
+ *
+ *  This code is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *  Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ */
+
+package jdk.incubator.foreign;
+
+import java.nio.ByteOrder;
+import java.util.Objects;
+import jdk.internal.vm.annotation.ForceInline;
+
+/**
+ * This class provides convenient methods for copying data between primitive arrays and memory segments.
+ *
+ * <p>If the source (destination) segment is actually a view of the destination (source) array,
+ * and if the copy region of the source overlaps with the copy region of the destination,
+ * the copy of the overlapping region is performed as if the data in the overlapping region
+ * were first copied into a temporary segment before being copied to the destination.</p>
+ *
+ * <p>
+ * Copy operations defined in this class accept a <em>byte order</em> parameter. If the specified byte order is different
+ * from the <em>native</em> byte order, a byte swap operation is performed on each array elements
```

Comment:
That seems consistent with MemoryAccess, to be honest.

Reply:
True. Ok, that actually makes sense now that I compare it with what memaccess var handles do. i.e. the byte order always corresponds to the memory segment, since the array is of course in native order. So, in practice this means a swap occurs when the specified or is non-native. Thanks.